auth.php

```diff
@@ -47,86 +47,44 @@
+    unset($SESSION->wantsurl);
     redirect(new moodle_url($wantsurl));
+} else if ($state == \tool_mfa\plugininfo\factor::STATE_FAIL) {
+    \tool_mfa\manager::cannot_login();
 }
 
-$userfactors = \tool_mfa\plugininfo\factor::get_active_user_factor_types();
-
-if (count($userfactors) > 0) {
-    $nextfactor = \tool_mfa\plugininfo\factor::get_next_user_factor();
-    $gracemode = false;
-    $factorname = $nextfactor ? $nextfactor->name : null;
-} else {
-    $factorname = null;
-    $gracemode = true;
-}
-
-$form = new login_form($currenturl, array('factor_name' => $factorname, 'grace_mode' => $gracemode));
-if (isset($nextfactor)) {
-    $factor = \tool_mfa\plugininfo\factor::get_factor($factorname);
+// If not pass/fail, check next factor state.
+$factor = \tool_mfa\plugininfo\factor::get_next_user_factor();
+if ($factor->get_state() == \tool_mfa\plugininfo\factor::STATE_FAIL) {
+    // Catch any instant fail state not caught by manager (fallback).
+    \tool_mfa\manager::cannot_login();
 }
 
+// If not, perform form actions for input factor.
+$form = new login_form($currenturl, array('factor' => $factor));
 if ($form->is_submitted()) {
     $form->is_validated();
 
+    // Set state from user actions.
     if ($form->is_cancelled()) {
-        if (isset($factor)) {
-            // Only set vars if from a factor.
-            $factor->set_state(\tool_mfa\plugininfo\factor::STATE_NEUTRAL);
-        }
+        $factor->set_state(\tool_mfa\plugininfo\factor::STATE_NEUTRAL);
     } else {
         if ($data = $form->get_data()) {
-            if (isset($factor)) {
-                $factor->set_state(\tool_mfa\plugininfo\factor::STATE_PASS);
-            }
-        }
-    }
-}
-
-if ($form->is_submitted()
-    && (isset($factor) && $factor->get_state() != \tool_mfa\plugininfo\factor::STATE_FAIL)
-    || !isset($factor)) {
-
-    if ($next = \tool_mfa\plugininfo\factor::get_next_user_factor()) {
-        // Fallback factor means there are not enough factors setup or answered. Requires special handling.
-        if ($next->name == 'fallback') {
-            tool_mfa_logout();
-            print_error('error:notenoughfactors', 'tool_mfa', new moodle_url('/'));
-        } else {
-            redirect($currenturl);
+            $factor->set_state(\tool_mfa\plugininfo\factor::STATE_PASS);
         }
     }
 
-    if (tool_mfa_user_passed_enough_factors() || $gracemode) {
-        $SESSION->tool_mfa_authenticated = true;
-
-        $event = \tool_mfa\event\user_passed_mfa::user_passed_mfa_event($USER);
-        $event->trigger();
-
-        if ($gracemode) {
-            redirect(new moodle_url('/admin/tool/mfa/user_preferences.php'));
-        }
-
-        if (!empty($SESSION->wantsurl)) {
-            unset($SESSION->wantsurl);
-        }
-        redirect(new moodle_url($wantsurl));
-    }
-
-    tool_mfa_logout();
-    print_error('error:notenoughfactors', 'tool_mfa', new moodle_url('/'));
+    // Move to next factor.
+    redirect($currenturl);
```

Comment:
At this point we may have just set the current factor to passed, and so now the whole thing is a pass. In which case it would be better to directly redirect to the wants urls, instead of reload this page which then redirects after check_status() at the top. At the same time we could also have a neutral which means we now fail overall and we should just show the error instead of redirecting and then showing the error.

So how about we just call check_status() here (when it is actually doing a check not a get) and you can pass it a boolean to see if you want it to redirect to auth.php or not.